src/main/scala/io/iohk/ethereum/blockchain/sync/regular/BlockBroadcast.scala

```diff
@@ -18,35 +20,39 @@ class BlockBroadcast(val etcPeerManager: ActorRef, syncConfig: SyncConfig) {
     * The hash of the block is sent to all of those peers while the block itself is only sent to
     * the square root of the total number of those peers, with the subset being obtained randomly.
     *
-    * @param newBlock, block to broadcast
+    * @param blockToBroadcast, block to broadcast
     * @param handshakedPeers, to which the blocks will be broadcasted to
     */
-  def broadcastBlock(newBlock: NewBlock, handshakedPeers: Map[Peer, PeerInfo]): Unit = {
-    val peersWithoutBlock = handshakedPeers.collect {
-      case (peer, peerInfo) if shouldSendNewBlock(newBlock, peerInfo) => peer
-    }.toSet
+  def broadcastBlock(blockToBroadcast: BlockToBroadcast, handshakedPeers: Map[Peer, PeerInfo]): Unit = {
+    val peersWithoutBlock = handshakedPeers.filter { case (_, peerInfo) =>
+      shouldSendNewBlock(blockToBroadcast, peerInfo)
+    }
 
-    broadcastNewBlock(newBlock, peersWithoutBlock)
+    broadcastNewBlock(blockToBroadcast, peersWithoutBlock)
 
     if (syncConfig.broadcastNewBlockHashes) {
       // NOTE: the usefulness of this message is debatable, especially in private networks
-      broadcastNewBlockHash(newBlock, peersWithoutBlock)
+      broadcastNewBlockHash(blockToBroadcast, peersWithoutBlock.keySet)
     }
   }
 
-  private def shouldSendNewBlock(newBlock: NewBlock, peerInfo: PeerInfo): Boolean =
+  private def shouldSendNewBlock(newBlock: BlockToBroadcast, peerInfo: PeerInfo): Boolean =
     newBlock.block.header.number > peerInfo.maxBlockNumber ||
       newBlock.chainWeight > peerInfo.chainWeight
 
-  private def broadcastNewBlock(newBlock: NewBlock, peers: Set[Peer]): Unit =
-    obtainRandomPeerSubset(peers).foreach { peer =>
-      etcPeerManager ! EtcPeerManagerActor.SendMessage(newBlock, peer.id)
+  private def broadcastNewBlock(blockToBroadcast: BlockToBroadcast, peers: Map[Peer, PeerInfo]): Unit =
+    obtainRandomPeerSubset(peers.keySet).foreach { peer =>
```

Comment:
@mirkoAlic, beware that `obtainRandomPeerSubset` always returns the same subset for the same `peers`!
Because `Set` is unordered collection, Random.shuffle does not work here, please check REPL:

```
scala> Set(1,2,3,4,5)
res0: scala.collection.immutable.Set[Int] = Set(5, 1, 2, 3, 4)

scala> util.Random.shuffle(res0)
res1: scala.collection.immutable.Set[Int] = Set(5, 1, 2, 3, 4)
```

Reply:
Excellent catch! Furthermore, this is an open issue we have in the dissemination of the new block  message in general.

Reply:
Great catch! I created a separate task to deal with it (https://jira.iohk.io/browse/ETCM-415) and will be working in parallel to speed it up

Reply:
Done: https://github.com/input-output-hk/mantis/pull/814